File: framework/languages/dotnet/e2e/RuntimeMonitoring/Server/Service/Handlers/ServiceHandlers.cs
using RuntimeMonitoring.Server.Service.Support;
using RuntimeMonitoring.Shared;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Handlers;
using Zlink.Framework.Contracts.Messaging;
using Zlink.Framework.Contracts.Spots;
using Zlink.Framework.Contracts.Timers;

namespace RuntimeMonitoring.Server.Service.Handlers;

internal sealed class ProfileRequestHandler(
    EvidenceStore evidence,
    ApplicationDispatchGate? applicationGate = null)
    : IZLinkRequestHandler<ProfileReq, ProfileRes>
{
    public async ValueTask<ProfileRes> HandleAsync(
        ProfileReq request,
        IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"profile-request|rid={evidence.Rid}|marker={request.Marker}|value={request.Value}");
        if (string.Equals(
                request.Value,
                "application-gate",
                StringComparison.Ordinal))
        {
            evidence.Add(
                $"application-gate-enter|rid={evidence.Rid}|marker={request.Marker}");
            await (applicationGate
                   ?? throw new InvalidOperationException(
                       "Application dispatch gate is not configured."))
                .WaitAsync(cancellationToken);
            evidence.Add(
                $"application-gate-exit|rid={evidence.Rid}|marker={request.Marker}");
        }
        return new ProfileRes(
            $"profile:{request.Value}",
            evidence.Rid,
            request.Marker);
    }
}

internal sealed class MonitoringActor(
    string actorId,
    IZLinkActorContext context) : IZLinkActor
{
    public string ActorId { get; } = actorId;

    public IZLinkActorContext Context { get; } = context;
}

internal sealed class MonitoringActorFactory : IZLinkActorFactory<MonitoringActor>
{
    public ValueTask<MonitoringActor> CreateAsync(
        IZLinkActorContext context,
        CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(new MonitoringActor(context.ActorId, context));
    }
}

internal sealed class MonitoringEntrySpot(IZLinkEntrySpotContext context)
    : IZLinkEntrySpot<MonitoringActor>
{
    public IZLinkEntrySpotContext Context { get; } = context;

    public ValueTask<ZLinkActorCreateResponse> OnCreateActorAsync(
        MonitoringActor actor,
        ZLinkMessage createRequest,
        CancellationToken cancellationToken)
    {
        _ = actor;
        _ = createRequest;
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(ZLinkActorCreateResponse.Accept());
    }

    public ValueTask<ZLinkSpotActorJoinResult> OnActorJoinAsync(
        string actorId,
        ZLinkMessage request,
        CancellationToken cancellationToken)
    {
        _ = actorId;
        _ = request;
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(ZLinkSpotActorJoinResult.Accept());
    }

    public ValueTask OnJoinedActorAsync(
        MonitoringActor actor,
        CancellationToken cancellationToken)
    {
        _ = actor;
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.CompletedTask;
    }

    public ValueTask OnLeaveActorAsync(
        MonitoringActor actor,
        CancellationToken cancellationToken)
    {
        _ = actor;
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.CompletedTask;
    }

    public async ValueTask OnInitializeAsync(CancellationToken cancellationToken)
    {
        await Context.AddTimer<FailingTimerHandler>(
            "failing",
            TimeSpan.FromMilliseconds(50),
            new ZLinkTimerOptions { StopOnUnhandledException = false },
            cancellationToken);
        await Context.AddTimer<FailingTimerHandler>(
            "stopping",
            TimeSpan.FromMilliseconds(50),
            new ZLinkTimerOptions { StopOnUnhandledException = true },
            cancellationToken);
    }
}

internal sealed class FailingTimerHandler : IZLinkSpotTimerHandler<MonitoringEntrySpot>
{
    public ValueTask HandleAsync(
        MonitoringEntrySpot spot,
        ZLinkTimerTick tick,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        throw new InvalidOperationException("monitoring timer failure");
    }
}

internal sealed class MonitoringSubjectSpot(IZLinkSpotContext context)
    : IZLinkSpot
{
    public IZLinkSpotContext Context { get; } = context;

    public void Configure()
    {
        Context.Handlers.AddSubscribe<MonitoringSubjectHandler>(
            RuntimeMonitoringNames.SpotChannel,
            "monitor.dynamic");
        Context.Handlers.AddSubscribe<MonitoringSubjectHandler>(
            RuntimeMonitoringNames.SpotChannel,
            "monitor.blocker");
    }

}

internal sealed class MonitoringSubjectHandler(
    EvidenceStore evidence,
    ApplicationDispatchGate gate)
    : IZLinkSpotSubscriptionHandler<MonitoringSubjectSpot, ProfileEvent>
{
    public async ValueTask HandleAsync(
        MonitoringSubjectSpot spot,
        ProfileEvent message,
        ZLinkPublishMessageContext context,
        CancellationToken cancellationToken)
    {
        _ = context;
        if (spot.Context.SpotId == "monitor-blocked")
        {
            evidence.Add(
                $"application-gate-enter|spot={spot.Context.SpotId}|topic={context.Topic}");
            await gate.WaitAsync(cancellationToken).ConfigureAwait(false);
        }
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add(
            $"logical-publish|spot={spot.Context.SpotId}|topic={context.Topic}|marker={message.Marker}");
    }
}

File: framework/languages/node/e2e/RuntimeMonitoring/Server/Service/Handlers/service-handlers.ts
import { Injectable } from '@nestjs/common';
import type {
  ZLinkActor,
  ZLinkActorCreateResponse,
  ZLinkActorContext,
  ZLinkEntrySpot,
  ZLinkEntrySpotContext,
  ZLinkMessage,
  ZLinkMessageContext,
  ZLinkPublishMessageContext,
  ZLinkRequestHandler,
  ZLinkSpot,
  ZLinkSpotContext,
  ZLinkSpotSubscriptionHandler,
  ZLinkSpotTimerHandler,
  ZLinkTimerTick
} from '@zlink-systems/framework';
import {
  zlinkEntrySpotSubscriptionHandler,
  zlinkSpotTimerHandler
} from '@zlink-systems/nestjs';
import { RuntimeMonitoringNames, type MonitoringPublish, type ProfileReq, type ProfileRes } from '../../../Shared/messages';
import { EvidenceStore } from '../Infrastructure/evidence-store';

@Injectable()
export class MonitoringPublishGate {
  private static readonly blocked = new Set<string>();
  private static readonly waiters = new Map<string, Set<() => void>>();

  setBlocked(target: string, blocked: boolean): void {
    if (blocked) {
      MonitoringPublishGate.blocked.add(target);
      return;
    }
    MonitoringPublishGate.blocked.delete(target);
    const pending = MonitoringPublishGate.waiters.get(target);
    MonitoringPublishGate.waiters.delete(target);
    for (const resolve of pending ?? []) resolve();
  }

  isBlocked(target: string): boolean {
    return MonitoringPublishGate.blocked.has(target);
  }

  async wait(target: string): Promise<void> {
    if (!MonitoringPublishGate.blocked.has(target)) return;
    await new Promise<void>((resolve) => {
      const pending = MonitoringPublishGate.waiters.get(target) ?? new Set<() => void>();
      pending.add(resolve);
      MonitoringPublishGate.waiters.set(target, pending);
    });
  }
}

@Injectable()
export class ProfileRequestHandler implements ZLinkRequestHandler<ProfileReq, ProfileRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: ProfileReq, context: ZLinkMessageContext): Promise<ProfileRes> {
    this.evidence.add(
      `profile-request|rid=${this.evidence.rid}|marker=${request.marker}`
      + `|value=${request.value}|packet=${context.packetName}`
    );
    return { value: `profile:${request.value}`, providerRid: this.evidence.rid, marker: request.marker };
  }
}

export class MonitoringActor implements ZLinkActor {
  constructor(
    readonly actorId: string,
    readonly context: ZLinkActorContext
  ) {}
}

export class MonitoringActorFactory {
  async create(context: ZLinkActorContext): Promise<MonitoringActor> {
    return new MonitoringActor(context.actorId, context);
  }
}

@Injectable()
@zlinkEntrySpotSubscriptionHandler({
  entrySpot: () => MonitoringEntrySpot,
  channelName: RuntimeMonitoringNames.spotChannel,
  topic: RuntimeMonitoringNames.publishTopic
})
export class MonitoringEntryPublishHandler
  implements ZLinkSpotSubscriptionHandler<MonitoringEntrySpot, MonitoringPublish> {
  constructor(
    private readonly evidence: EvidenceStore,
    private readonly gate: MonitoringPublishGate
  ) {}

  async handle(
    spot: MonitoringEntrySpot,
    event: MonitoringPublish,
    context: ZLinkPublishMessageContext
  ): Promise<void> {
    void context;
    await this.gate.wait(`entry:${this.evidence.rid}`);
    this.evidence.add(
      `publish-received|rid=${this.evidence.rid}|spot=entry|marker=${event.marker}`
    );
    void spot;
  }
}

export class MonitoringUserSpot implements ZLinkSpot<MonitoringActor> {
  readonly context!: ZLinkSpotContext<MonitoringActor, MonitoringUserSpot>;

  configure(): void {
    this.context.handlers.addSubscribe(
      MonitoringUserSpotPublishHandler,
      RuntimeMonitoringNames.spotChannel,
      RuntimeMonitoringNames.publishTopic
    );
    this.requireEvidence().add(
      `subscription-configured|rid=${this.requireEvidence().rid}|spot=${this.context.spotId}`
    );
  }

  async onInitialize(): Promise<void> {
    this.requireEvidence().add(`spot-ready|rid=${this.requireEvidence().rid}|spot=${this.context.spotId}`);
  }

  async onActorJoin(_actorId: string, _request: ZLinkMessage): Promise<{ accepted: boolean }> {
    return { accepted: true };
  }

  async onJoinedActor(_actor: MonitoringActor): Promise<void> {}

  async onLeaveActor(_actor: MonitoringActor): Promise<void> {}

  async onDisconnectActor(_actor: MonitoringActor): Promise<void> {}

  private requireEvidence(): EvidenceStore {
    const evidence = MonitoringUserSpot.evidence;
    if (evidence === undefined) throw new Error('MonitoringUserSpot evidence is not configured.');
    return evidence;
  }

  private static evidence?: EvidenceStore;

  static useEvidence(evidence: EvidenceStore): void {
    MonitoringUserSpot.evidence = evidence;
  }
}

@Injectable()
export class MonitoringUserSpotPublishHandler
  implements ZLinkSpotSubscriptionHandler<MonitoringUserSpot, MonitoringPublish> {
  constructor(
    private readonly evidence: EvidenceStore,
    private readonly gate: MonitoringPublishGate
  ) {}

  async handle(
    spot: MonitoringUserSpot,
    event: MonitoringPublish,
    context: ZLinkPublishMessageContext
  ): Promise<void> {
    void context;
    const spotId = String(spot.context.spotId);
    const payloadBytes = event.blocker?.length ?? 0;
    this.evidence.add(
      `publish-entered|rid=${this.evidence.rid}|spot=${spotId}|marker=${event.marker}|payloadBytes=${payloadBytes}`
    );
    await this.gate.wait(`spot:${spotId}`);
    this.evidence.add(
      `publish-received|rid=${this.evidence.rid}|spot=${spotId}|marker=${event.marker}|payloadBytes=${payloadBytes}`
    );
  }
}

@Injectable()
export class MonitoringEntrySpot implements ZLinkEntrySpot<MonitoringActor> {
  declare readonly context: ZLinkEntrySpotContext<MonitoringActor>;

  async onJoinedActor(_actor: ZLinkActor): Promise<void> {}

  async onLeaveActor(_actor: ZLinkActor): Promise<void> {}

  async onDisconnectActor(_actor: ZLinkActor): Promise<void> {}

  async onInitialize(): Promise<void> {
    await this.context.addTimer('failing', 1000, FailingTimerHandler, { stopOnUnhandledException: false });
    await this.context.addTimer('stopping', 1000, FailingTimerHandler, { stopOnUnhandledException: true });
  }

  async onCreateActor(_actor: MonitoringActor, _createRequest: ZLinkMessage): Promise<ZLinkActorCreateResponse> {
    return { accepted: true };
  }
}

@Injectable()
@zlinkSpotTimerHandler()
export class FailingTimerHandler implements ZLinkSpotTimerHandler<ZLinkEntrySpot> {
  private readonly failedTimers = new Set<string>();

  async handle(_spot: ZLinkEntrySpot, tick: ZLinkTimerTick): Promise<void> {
    if (this.failedTimers.has(tick.name)) return;
    this.failedTimers.add(tick.name);
    throw new Error('monitoring timer failure');
  }
}
